xtask/src/ui.rs

```diff
@@ -0,0 +1,480 @@
+use std::path::{Path, PathBuf};
+use std::process::{Command, ExitStatus, Output, Stdio};
+use std::{env, fs};
+
+use crate::{temp, util, ImageTarget};
+use clap::Args;
+use cross::shell::MessageInfo;
+use cross::{docker, CommandExt, ToUtf8};
+
+#[derive(Args, Debug)]
+pub struct UiTest {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    pub verbose: bool,
+    /// Do not print cross log messages.
+    #[clap(short, long)]
+    pub quiet: bool,
+    /// Coloring: auto, always, never
+    #[clap(long)]
+    pub color: Option<String>,
+    /// Container engine (such as docker or podman).
+    #[clap(long)]
+    pub engine: Option<String>,
+    /// Specify a tag to use instead of the derived one, eg `local`
+    #[clap(long)]
+    pub tag: Option<String>,
+    /// Repository name for image.
+    #[clap(long, default_value = docker::CROSS_IMAGE)]
+    pub repository: String,
+    /// Do not pull the image (local image).
+    #[clap(long)]
+    pub no_pull: bool,
+    /// Target to run tests for.
+    #[clap(long)]
+    pub target: Option<ImageTarget>,
+}
+
+struct Utf8Output {
+    command: String,
+    status: ExitStatus,
+    stdout: String,
+    stderr: String,
+}
+
+impl Utf8Output {
+    fn bail(&self, message: &str) -> eyre::ErrReport {
+        eyre::eyre!(
+            r#"command unexpectedly {message} with exit status of {:?}:
+command
+=======
+{}
+
+stdout
+======
+{}
+
+stderr
+======
+{}
+"#,
```

Comment:
This is just so we get a very pretty diagnostic output of the actual error/output messages, the command, and the exit status if anything fails our expectations. This makes debugging way easier.